storage/innobase/srv/srv0srv.cc

```diff
@@ -1734,9 +1790,40 @@ static uint32_t srv_do_purge(ulint* n_total_purged)
 
 	if (n_use_threads == 0) {
 		n_use_threads = n_threads;
+		compute_series(series, innodb_purge_threads_MAX + 1,
+			(100 - (adaptive_purge_threshold + safety_net)),
+			n_threads);
 	}
 
+	/* user can configure > 1 purge threads. purge operation generates
+	redo records. user workload too generates redo records. this dual redo
+	generation can cause redo log to hit the threshold there-by increasing
+	performance jitter, especially with user visible workload.
+	adaptive purge logic abates this effect by dynamically adjusting
+	purge threads (in turn the redo-log generation) based on redo-log fill
+	factor. This greatly helps stabilize and also improve overall
+        performance of user visible workload at a slight cost of increasing
+	history-length.
+	Continous growth of history length could increase select latency.
+	To handle this effect, adaptive logic continue to increase and decrease
+	purge threads there-by keeping a check on history-length without
+	any visible impact on performance (infact sysbench update-index
+	which involves point-select has shown significant performance
+	improvement) */
+	static lsn_t lsn_lwm = adaptive_purge_threshold;
+	static lsn_t lsn_hwm = adaptive_purge_threshold + series[n_use_threads];
+	static ulint start_time = 0;
+	static lsn_t lsn_age_factor = get_lsn_age_factor();
+
 	do {
+		if ((ut_time_ms() - start_time) > 1000) {
+			/* refresh lsn_age_factor once per second to avoid
+			too frequent purge threads increase/decrease and
+			flush list mutex contention */
+			lsn_age_factor = get_lsn_age_factor();
+			start_time = ut_time_ms();
+		}
```

Comment:
I would suggest to use `my_interval_timer()` directly instead of the wrapper. But, instead of polling the clock, could we simply ensure that `purge_coordinator_callback()` is scheduled at periodic intervals? Maybe we could have a dedicated thread for the coordinator, similar to the dedicated `buf_flush_page_cleaner` that is normally triggered once per second?

In any case, it should suffice to read the clock once. The call `get_lsn_age_factor()` should not consume milliseconds unless the `buf_pool.flush_list_mutex` is extremely contended.

Reply:
1. my_interval_timer. done. save us a division.
2. as per my understanding, purge_cordinator continues to work till there is work to do and re-scheduled using purge_cordinator_timer. if there is something to improve in this logic we can surely consider it as a separate enhancement.
3. no more buf_pool.flush_list_mutex. (replaced with log-sys but should be quick as we read-only global variables).